CI Failure: SchedV2Test.test_moveVersions

```
com.ichi2.libanki.sched.SchedV2Test > test_moveVersions FAILED
    java.lang.Exception: Main looper has queued unexecuted runnables. This might be the cause of the test failure. You might need a shadowOf(getMainLooper()).idle() call.
        at org.robolectric.android.internal.AndroidTestEnvironment.checkStateAfterTestFailure(AndroidTestEnvironment.java:470)
        at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:548)
        at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:252)
        at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
        Caused by:
        java.lang.AssertionError: expected:<50> but was:<51>
            at org.junit.Assert.fail(Assert.java:89)
            at org.junit.Assert.failNotEquals(Assert.java:835)
            at org.junit.Assert.assertEquals(Assert.java:647)
            at org.junit.Assert.assertEquals(Assert.java:633)
            at com.ichi2.libanki.sched.SchedV2Test.test_moveVersions(SchedV2Test.java:1564)
```

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
@@ -226,7 +226,8 @@ public class Collection {
             throw new RuntimeException("Unsupported scheduler version");
         }
         modSchema();
-        SchedV2 v2Sched = new SchedV2(this);
+        @SuppressLint("VisibleForTests")
+        SchedV2 v2Sched = new SchedV2(this, mSched.getTime());
         clearUndo();
         if (ver == 1) {
             v2Sched.moveToV1();
@@ -2074,7 +2075,7 @@ public class Collection {
     public void replaceSchedulerForTests(Time time) {
         int ver = schedVer();
         if (ver == 1) {
-            throw new IllegalStateException("Not Implemented");
+            mSched = new Sched(this, time);
         } else if (ver == 2) {
             mSched = new SchedV2(this, time);
         }

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
@@ -1288,25 +1288,7 @@ public class Collection {
         Card clonedCard = card.clone();
         Undoable undoableReview = new Undoable(REVIEW) {
             public long undo(Collection col) {
-                // remove leech tag if it didn't have it before
-                if (!wasLeech && clonedCard.note().hasTag("leech")) {
-                    clonedCard.note().delTag("leech");
-                    clonedCard.note().flush();
-                }
-                Timber.i("Undo Review of card %d, leech: %b", clonedCard.getId(), wasLeech);
-                // write old data
-                clonedCard.flush(false);
-                // and delete revlog entry
-                long last = col.getDb().queryLongScalar("SELECT id FROM revlog WHERE cid = ? ORDER BY id DESC LIMIT 1", new Object[] {clonedCard.getId()});
-                col.getDb().execute("DELETE FROM revlog WHERE id = " + last);
-                // restore any siblings
-                col.getDb().execute("update cards set queue=type,mod=?,usn=? where queue=" + Consts.QUEUE_TYPE_SIBLING_BURIED + " and nid=?",
-                        new Object[] {Utils.intTime(), col.usn(), clonedCard.getNid()});
-                // and finally, update daily count
-                @Consts.CARD_QUEUE int n = clonedCard.getQueue() == Consts.QUEUE_TYPE_DAY_LEARN_RELEARN ? Consts.QUEUE_TYPE_LRN : clonedCard.getQueue();
-                String type = (new String[]{"new", "lrn", "rev"})[n];
-                col.getSched()._updateStats(clonedCard, type, -1);
-                col.getSched().setReps(col.getSched().getReps() - 1);
+                col.getSched().undoReview(clonedCard, wasLeech);
                 return clonedCard.getId();
             }
         };

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/sched/AbstractSched.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/sched/AbstractSched.java
@@ -186,6 +186,13 @@ public abstract class AbstractSched {
     public abstract int[] recalculateCounts();
     public abstract void setReportLimit(int reportLimit);
 
+    /**
+     * Reverts answering a card.
+     * 
+     * @param card The data of the card before the review was made
+     * @param wasLeech Whether the card was a leech before the review was made (if false, remove the leech tag)
+     * */
+    public abstract void undoReview(@NonNull Card card, boolean wasLeech);
 
     public interface LimitMethod {
         int operation(Deck g);

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/sched/AbstractSched.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/sched/AbstractSched.java
@@ -14,6 +14,7 @@ import com.ichi2.libanki.Consts;
 import com.ichi2.libanki.Deck;
 import com.ichi2.libanki.DeckConfig;
 import com.ichi2.libanki.Collection;
+import com.ichi2.libanki.utils.Time;
 
 import java.lang.ref.WeakReference;
 import java.util.List;
@@ -193,6 +194,14 @@ public abstract class AbstractSched {
      * @param wasLeech Whether the card was a leech before the review was made (if false, remove the leech tag)
      * */
     public abstract void undoReview(@NonNull Card card, boolean wasLeech);
+
+
+    /**
+     * @return The current time dependency that the scheduler is using (used for Unit Testing).
+     */
+    @NonNull
+    public abstract Time getTime();
+
 
     public interface LimitMethod {
         int operation(Deck g);

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java
@@ -39,6 +39,7 @@ import com.ichi2.libanki.Utils;
 import com.ichi2.libanki.Deck;
 import com.ichi2.libanki.DeckConfig;
 
+import com.ichi2.libanki.utils.Time;
 import com.ichi2.utils.Assert;
 import com.ichi2.utils.JSONArray;
 import com.ichi2.utils.JSONException;
@@ -92,6 +93,10 @@ public class Sched extends SchedV2 {
         super(col);
     }
 
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    public Sched(@NonNull Collection col, @NonNull Time time) {
+        super(col, time);
+    }
 
     @Override
     public void answerCard(Card card, @Consts.BUTTON_TYPE int ease) {
@@ -134,7 +139,7 @@ public class Sched extends SchedV2 {
             throw new RuntimeException("Invalid queue");
         }
         _updateStats(card, "time", card.timeTaken());
-        card.setMod(Utils.intTime());
+        card.setMod(mTime.intTime());
         card.setUsn(mCol.usn());
         card.flushSched();
     }
@@ -203,7 +208,7 @@ public class Sched extends SchedV2 {
         String sids = Utils.ids2str(allDecks);
         mCol.log(mCol.getDb().queryLongList("select id from cards where " + queueIsBuriedSnippet() + " and did in " + sids));
         mCol.getDb().execute("update cards set mod=?,usn=?," + _restoreQueueSnippet() + " where " + queueIsBuriedSnippet() + " and did in " + sids,
-                Utils.intTime(), mCol.usn());
+                mTime.intTime(), mCol.usn());
     }
 
     /**
@@ -405,7 +410,7 @@ public class Sched extends SchedV2 {
     @Override
     protected Card _getLrnCard(boolean collapse) {
         if (_fillLrn()) {
-            double cutoff = Utils.now();
+            double cutoff = mTime.now();
             if (collapse) {
                 cutoff += mCol.getConf().getInt("collapseTime");
             }
@@ -467,11 +472,11 @@ public class Sched extends SchedV2 {
                 }
             }
             int delay = _delayForGrade(conf, card.getLeft());
-            if (card.getDue() < Utils.now()) {
+            if (card.getDue() < mTime.now()) {
                 // not collapsed; add some randomness
                 delay *= Utils.randomFloatInRange(1f, 1.25f);
             }
-            card.setDue((int) (Utils.now() + delay));
+            card.setDue((int) (mTime.now() + delay));
 
             // due today?
             if (card.getDue() < mDayCutoff) {
@@ -601,7 +606,7 @@ public class Sched extends SchedV2 {
     private void log(long id, int usn, @Consts.BUTTON_TYPE int ease, int ivl, int lastIvl, int factor, int timeTaken, @Consts.REVLOG_TYPE int type) {
         try {
             mCol.getDb().execute("INSERT INTO revlog VALUES (?,?,?,?,?,?,?,?,?)",
-                    Utils.now() * 1000, id, usn, ease, ivl, lastIvl, factor, timeTaken, type);
+                    mTime.now() * 1000, id, usn, ease, ivl, lastIvl, factor, timeTaken, type);
         } catch (SQLiteConstraintException e) {
             try {
                 Thread.sleep(10);
@@ -631,7 +636,7 @@ public class Sched extends SchedV2 {
         mCol.getDb().execute(
                 "update cards set due = odue, queue = " + Consts.QUEUE_TYPE_REV + ", mod = ?" +
                 ", usn = ?, odue = 0 where queue IN (" + Consts.QUEUE_TYPE_LRN + "," + Consts.QUEUE_TYPE_DAY_LEARN_RELEARN + ") and type = " + Consts.CARD_TYPE_REV + " " + extra,
-                Utils.intTime(), mCol.usn());
+                mTime.intTime(), mCol.usn());
         // new cards in learning
         forgetCards(Utils.collection2Array(mCol.getDb().queryLongList( "SELECT id FROM cards WHERE queue IN (" + Consts.QUEUE_TYPE_LRN + "," + Consts.QUEUE_TYPE_DAY_LEARN_RELEARN + ") " + extra)));
     }
@@ -642,7 +647,7 @@ public class Sched extends SchedV2 {
                     "SELECT sum(left / 1000) FROM (SELECT left FROM cards WHERE did = ?"
                             + " AND queue = " + Consts.QUEUE_TYPE_LRN + " AND due < ?"
                             + " LIMIT ?)",
-                    did, (Utils.intTime() + mCol.getConf().getInt("collapseTime")), mReportLimit);
+                    did, (mTime.intTime() + mCol.getConf().getInt("collapseTime")), mReportLimit);
             return cnt + mCol.getDb().queryScalar(
                     "SELECT count() FROM (SELECT 1 FROM cards WHERE did = ?"
                             + " AND queue = " + Consts.QUEUE_TYPE_DAY_LEARN_RELEARN + " AND due <= ?"
@@ -826,7 +831,7 @@ public class Sched extends SchedV2 {
             card.setODue(card.getDue());
         }
         delay = _delayForGrade(conf, 0);
-        card.setDue((long) (delay + Utils.now()));
+        card.setDue((long) (delay + mTime.now()));
         card.setLeft(_startingLeft(card));
         // queue 1
         if (card.getDue() < mDayCutoff) {
@@ -1056,7 +1061,7 @@ public class Sched extends SchedV2 {
 
     private void _moveToDyn(long did, List<Long> ids) {
         ArrayList<Object[]> data = new ArrayList<>();
-        //long t = Utils.intTime(); // unused variable present (and unused) upstream
+        //long t = mTime.intTime(); // unused variable present (and unused) upstream
         int u = mCol.usn();
         for (long c = 0; c < ids.size(); c++) {
             // start at -100000 so that reviews are all due
@@ -1208,7 +1213,7 @@ public class Sched extends SchedV2 {
     protected void _updateCutoff() {
         Integer oldToday = mToday;
         // days since col created
-        mToday = (int) ((Utils.now() - mCol.getCrt()) / 86400);
+        mToday = (int) ((mTime.now() - mCol.getCrt()) / 86400);
         // end of day cutoff
         mDayCutoff = mCol.getCrt() + ((mToday + 1) * 86400);
         if (oldToday != mToday) {
@@ -1344,7 +1349,7 @@ public class Sched extends SchedV2 {
         mCol.getDb().execute(
                 "UPDATE cards SET queue = " + Consts.QUEUE_TYPE_SUSPENDED + ", mod = ?, usn = ? WHERE id IN "
                         + Utils.ids2str(ids),
-                Utils.intTime(), mCol.usn());
+                mTime.intTime(), mCol.usn());
     }
 
     protected String queueIsBuriedSnippet() {
@@ -1364,7 +1369,7 @@ public class Sched extends SchedV2 {
         mCol.getDb().execute(
                 "UPDATE cards SET " + _restoreQueueSnippet() + ", mod = ?, usn = ?"
                         + " WHERE queue = " + Consts.QUEUE_TYPE_SUSPENDED + " AND id IN " + Utils.ids2str(ids),
-                Utils.intTime(), mCol.usn());
+                mTime.intTime(), mCol.usn());
     }
 
 
@@ -1380,7 +1385,7 @@ public class Sched extends SchedV2 {
         remFromDyn(cids);
         removeLrn(cids);
         mCol.getDb().execute("update cards set " + queueIsBuriedSnippet() + ",mod=?,usn=? where id in " + Utils.ids2str(cids),
-                Utils.now(), mCol.usn());
+                mTime.now(), mCol.usn());
     }
 
     /**
@@ -1391,7 +1396,7 @@ public class Sched extends SchedV2 {
     @Override
     public void sortCards(long[] cids, int start, int step, boolean shuffle, boolean shift) {
         String scids = Utils.ids2str(cids);
-        long now = Utils.intTime();
+        long now = mTime.intTime();
         ArrayList<Long> nids = new ArrayList<>();
         for (long id : cids) {
             long nid = mCol.getDb().queryLongScalar("SELECT nid FROM cards WHERE id = ?",

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java
@@ -2894,6 +2894,13 @@ public class SchedV2 extends AbstractSched {
     }
 
 
+    @NonNull
+    @Override
+    public Time getTime() {
+        return mTime;
+    }
+
+
     /** End #5666 */
     public void discardCurrentCard() {
         mCurrentCard = null;

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java
@@ -2870,6 +2870,30 @@ public class SchedV2 extends AbstractSched {
         this.mReportLimit = reportLimit;
     }
 
+    @Override
+    public void undoReview(@NonNull Card oldCardData, boolean wasLeech) {
+        // remove leech tag if it didn't have it before
+        if (!wasLeech && oldCardData.note().hasTag("leech")) {
+            oldCardData.note().delTag("leech");
+            oldCardData.note().flush();
+        }
+        Timber.i("Undo Review of card %d, leech: %b", oldCardData.getId(), wasLeech);
+        // write old data
+        oldCardData.flush(false);
+        // and delete revlog entry
+        long last = mCol.getDb().queryLongScalar("SELECT id FROM revlog WHERE cid = ? ORDER BY id DESC LIMIT 1", new Object[] {oldCardData.getId()});
+        mCol.getDb().execute("DELETE FROM revlog WHERE id = " + last);
+        // restore any siblings
+        mCol.getDb().execute("update cards set queue=type,mod=?,usn=? where queue=" + Consts.QUEUE_TYPE_SIBLING_BURIED + " and nid=?",
+                new Object[] {Utils.intTime(), mCol.usn(), oldCardData.getNid()});
+        // and finally, update daily count
+        @Consts.CARD_QUEUE int n = oldCardData.getQueue() == Consts.QUEUE_TYPE_DAY_LEARN_RELEARN ? Consts.QUEUE_TYPE_LRN : oldCardData.getQueue();
+        String type = (new String[]{"new", "lrn", "rev"})[n];
+        _updateStats(oldCardData, type, -1);
+        setReps(getReps() - 1);
+    }
+
+
     /** End #5666 */
     public void discardCurrentCard() {
         mCurrentCard = null;

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java
@@ -151,7 +151,7 @@ public class SchedV2 extends AbstractSched {
      */
 
     // Not in libAnki.
-    private final Time mTime;
+    protected final Time mTime;
 
     /**
      * card types: 0=new, 1=lrn, 2=rev, 3=relrn
@@ -2885,7 +2885,7 @@ public class SchedV2 extends AbstractSched {
         mCol.getDb().execute("DELETE FROM revlog WHERE id = " + last);
         // restore any siblings
         mCol.getDb().execute("update cards set queue=type,mod=?,usn=? where queue=" + Consts.QUEUE_TYPE_SIBLING_BURIED + " and nid=?",
-                new Object[] {Utils.intTime(), mCol.usn(), oldCardData.getNid()});
+                new Object[] {mTime.intTime(), mCol.usn(), oldCardData.getNid()});
         // and finally, update daily count
         @Consts.CARD_QUEUE int n = oldCardData.getQueue() == Consts.QUEUE_TYPE_DAY_LEARN_RELEARN ? Consts.QUEUE_TYPE_LRN : oldCardData.getQueue();
         String type = (new String[]{"new", "lrn", "rev"})[n];

--- a/AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedV2Test.java
+++ b/AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedV2Test.java
@@ -19,6 +19,7 @@ package com.ichi2.libanki.sched;
 import android.database.Cursor;
 
 import com.ichi2.anki.RobolectricTest;
+import com.ichi2.anki.exception.ConfirmModSchemaException;
 import com.ichi2.libanki.Card;
 import com.ichi2.libanki.Collection;
 import com.ichi2.libanki.Consts;
@@ -204,10 +205,14 @@ public class SchedV2Test extends RobolectricTest {
 
     public Collection getColV2() throws Exception {
         Collection col = getCol();
-        col.changeSchedulerVer(2);
+        changeSchedulerVer(col, 2);
+        return col;
+    }
+
+    private void changeSchedulerVer(Collection col, int ver) throws ConfirmModSchemaException {
+        col.changeSchedulerVer(ver);
         col.setCrt(1596540138L);
         col.replaceSchedulerForTests(mTime);
-        return col;
     }
 
     private double now() {
@@ -1509,7 +1514,7 @@ public class SchedV2Test extends RobolectricTest {
     @Test
     public void test_moveVersions() throws Exception {
         Collection col = getColV2();
-        col.changeSchedulerVer(1);
+        changeSchedulerVer(col, 1);
 
         Note n = col.newNote();
         n.setItem("Front", "one");
@@ -1521,7 +1526,7 @@ public class SchedV2Test extends RobolectricTest {
         col.getSched().answerCard(c, 1);
 
         // the move to v2 should reset it to new
-        col.changeSchedulerVer(2);
+        changeSchedulerVer(col, 2);
         c.load();
         assertEquals(QUEUE_TYPE_NEW, c.getQueue());
         assertEquals(CARD_TYPE_NEW, c.getType());
@@ -1535,7 +1540,7 @@ public class SchedV2Test extends RobolectricTest {
         assertEquals(QUEUE_TYPE_MANUALLY_BURIED, c.getQueue());
 
         // revert to version 1
-        col.changeSchedulerVer(1);
+        changeSchedulerVer(col, 1);
 
         // card should have moved queues
         c.load();
@@ -1548,7 +1553,7 @@ public class SchedV2Test extends RobolectricTest {
         assertEquals(QUEUE_TYPE_NEW, c.getType());
 
         // make sure relearning cards transition correctly to v1
-        col.changeSchedulerVer(2);
+        changeSchedulerVer(col, 2);
         // card with 100 day interval, answering again
         col.getSched().reschedCards(new long[] {c.getId()}, 100, 100);
         c.load();
@@ -1561,9 +1566,9 @@ public class SchedV2Test extends RobolectricTest {
         c = col.getSched().getCard();
         col.getSched().answerCard(c, 1);
         // due should be correctly set when removed from learning early
-        col.changeSchedulerVer(1);
+        changeSchedulerVer(col, 1);
         c.load();
-        assumeThat("Issue 6814 - flaky test", c.getDue(), is(50));
+        assertEquals(50, c.getDue());
     }
 
 

